lightpath/controller.py

```diff
@@ -42,115 +50,403 @@ def __init__(self, client, endstations=None):
+            for src in self.sources:
+                if nx.has_path(self.graph, src, branch):
+                    found_paths = nx.all_simple_paths(self.graph,
+                                                      source=src,
+                                                      target=branch)
+                    paths.extend(found_paths)
+                else:
+                    logger.debug(f'No path between {src} and {branch}')
+
+        self.beamlines[endstation] = []
+        for path in paths:
+            subgraph = self.graph.subgraph(path)
+            devices = [node[1]['dev'] for node in subgraph.nodes.data()
+                       if node[1]['dev'] is not None]
+            bp = BeamPath(*devices, name=endstation)
+            self.beamlines[endstation].append(bp)
+
+        # This isn't used currently, but left for now
+        setattr(self, bp.name.replace(' ', '_').lower(),
+                self.beamlines[endstation])
+
+    def imped_z(self, path: BeamPath) -> float:
         """
-        Current device destinations for the LCLS photon beam
+        Get z position of impediment or inf.
+
+        Parameters
+        ----------
+        path : :class:`BeamPath`
+            BeamPath to find impediment position in
+
+        Returns
+        -------
+        float
+            z position of impediment
+        """
+        return getattr(path.impediment, 'md.z', math.inf)
+
+    def active_path(self, dest: str) -> BeamPath:
+        """
+        Return the most active path to the requested endstation
+
+        Looks for the path with the latest impediment (highest z)
+
+        Parameters
+        ----------
+        dest : str
+            endstation to look for paths towards
+
+        Returns
+        -------
+        path : :class:`BeamPath`
+            the active path
+        """
+        paths = self.beamlines[dest]
+        if len(paths) == 1:
+            return paths[0]
+
+        paths_by_length = sorted(paths, key=self.imped_z)
+
+        return paths_by_length[-1]
+
+    def walk_facility(self) -> Dict[NodeName, List[NodeName]]:
         """
-        return list(set([p.impediment for p in self.beamlines.values()
-                         if p.impediment and p.impediment not in p.branches]))
+        Return the paths from source to destination by walking the
+        graph along device destinations
+
+        Returns
+        -------
+        Dict[NodeName, List[NodeName]]
+            A mapping from source node names to path of nodes
+
+        Raises
+        ------
+        PathError
+            If a single, valid path cannot be determined
+        """
+        sources = [n for n in self.graph.nodes if 'source' in n]
+
+        paths: Dict[NodeName, List] = {k: [] for k in sources}
+
+        for src, path in paths.items():
+            successors = list(self.graph.successors(src))
+            # skip to node after source node
+            curr = successors[0]
```

Comment:
I believe there's some protection against that particular case since we only add to sources if a subgraph with that branch name exists.  Still good to add an exception catch here though